src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -0,0 +1,823 @@
+    MTRAsyncCallbackReadyHandler readyHandler = ^(MTRDevice * device, NSUInteger retryCount) {
+        [self connectAndPerformAsync:^(MTRBaseDevice * baseDevice) {
+            [baseDevice readAttributeWithEndpointId:endpointId
+                                          clusterId:clusterId
+                                        attributeId:attributeId
+                                             params:params
+                                        clientQueue:self.queue
+                                         completion:^(NSArray<NSDictionary<NSString *, id> *> * _Nullable values,
+                                             NSError * _Nullable error) {
+                                             if (values) {
+                                                 // Since the format is the same data-value dictionary, this looks like an attribute
+                                                 // report
+                                                 [self _handleAttributeReport:values];
+                                             }
+                                             // TODO: retry on error
+                                             [workItem endWork];
+                                         }];
+        }];
+    };
+    workItem.readyHandler = readyHandler;
+    [_asyncCallbackWorkQueue enqueueWorkItem:workItem];
+
+    // Return current known / expected value right away
+    MTRAttributePath * attributePath = [MTRAttributePath attributePathWithEndpointId:endpointId
+                                                                           clusterId:clusterId
+                                                                         attributeId:attributeId];
+    NSDictionary<NSString *, id> * attributeValueToReturn = [self _attributeValueDictionaryForAttributePath:attributePath];
+
+    return attributeValueToReturn;
+}
+
+- (void)writeAttributeWithEndpointId:(NSNumber *)endpointId
+                           clusterId:(NSNumber *)clusterId
+                         attributeId:(NSNumber *)attributeId
+                               value:(id)value
+               expectedValueInterval:(NSNumber *)expectedValueIntervalMs
+                   timedWriteTimeout:(NSNumber * _Nullable)timeoutMs
+{
+    // Start the asynchronous operation
+    [self connectAndPerformAsync:^(MTRBaseDevice * baseDevice) {
+        [baseDevice
+            writeAttributeWithEndpointId:endpointId
+                               clusterId:clusterId
+                             attributeId:attributeId
+                                   value:value
+                       timedWriteTimeout:timeoutMs
+                             clientQueue:self.queue
+                              completion:^(NSArray<NSDictionary<NSString *, id> *> * _Nullable values, NSError * _Nullable error) {
+                                  if (values) {
+                                      [self _handleAttributeReport:values];
```

Comment:
This seems _very_ strange, since the values here will never be the sort of data values we expect in an attribute report....